### PR TITLE
Feature/view only account

### PR DESCRIPTION
**Pull request by Shramp**
TODO: update gitbook

Add view-only-accounts. These accounts are created by importing a private-view-key. They can view transactions for that view key and that's pretty much it. They'll be useful in hot + cold wallet flows and for viewing proof of burn.

Add view-only account functionality:
- add new tables for `view_only_accounts` and `view_only_txos`
- add `ViewOnlyAccount`, and `ViewOnlyTxo` models
- add `ViewOnlyAccount`, and `ViewOnlyTxo` services
- add `ViewOnlyBalance` to balance service
- add `ViewOnlyAccount` , `ViewOnlySecrets`, and `ViewOnlyTxo` json conversions
- update `sync.rs` to also sync view-only-accounts in a separate loop
- add endpoints:
  - export_view_only_account_secrets
  - get_view_only_account
  - get_all_view_only_accounts
  - get_balance_for_view_only_account
  - get_txos_for_view_only_account
  - import_view_only_account
  - remove_view_only_account
  - update_view_only_account_name

Soundtrack of this PR: [link to song that really fits the mood of this PR]()

### Motivation

< The motivation for the changes in this PR. "Currently we...", "This is needed because..." >

### In this PR
* < Additions, removals, fixes, features >

< Ticket status, e.g. "fixes #issue number" > 

### Future Work
* < Out of scope non-goals for this PR >
* < These should be links to tickets. If the tickets do not exist, make them. >

### CODE REVIEW

File: full-service/src/db/view_only_account.rs
Comment (remoun) [inline full-service/src/db/view_only_account.rs:146]: Optional/Equivalent: Add a top-level `use crate::db::schema` then these in-method declarations can be `use schema::...`

File: full-service/src/json_rpc/e2e.rs
Comment (remoun) [inline full-service/src/json_rpc/e2e.rs:3735]: Nit: This nested `json!` is unnecessary, judging by other nested usages in this file

File: full-service/src/json_rpc/wallet.rs
Comment (remoun) [inline full-service/src/json_rpc/wallet.rs:441]: Optional/Equivalent: `(&account).try_into()?` or maybe even `account.try_into().map_err(format_error)?`

File: full-service/src/json_rpc/wallet.rs
Comment (remoun) [inline full-service/src/json_rpc/wallet.rs:872]: IIUC these should use `format_invalid_request_error`

File: full-service/src/service/account.rs
Comment (remoun) [inline full-service/src/service/account.rs:50]: Nit: Colon before error message
```suggestion
    /// Error decoding base64: {0}
```

File: full-service/src/service/account.rs
Comment (remoun) [inline full-service/src/service/account.rs:53]: Ditto
```suggestion
    /// Error decoding private view key: {0}
```

File: full-service/src/service/sync.rs
Comment (remoun) [inline full-service/src/service/sync.rs:611]: Nit: Remove dead code

File: full-service/src/service/sync.rs
Comment (remoun) [inline full-service/src/service/sync.rs:641]: Optional/Equivalent: Use [`Range.map`](https://doc.rust-lang.org/std/ops/struct.Range.html#method.map) to DRY these calls.
```suggestion
            &(0..15).map(|_| o.clone()).collect::<Vec<_>>(),
```

File: full-service/src/service/view_only_account.rs
Comment (remoun) [inline full-service/src/service/view_only_account.rs:177]: Nit: Remove empty line
```suggestion
```

File: full-service/src/db/view_only_account.rs
Comment (remoun) [inline full-service/src/db/view_only_account.rs:1]: Nit: Use `-2022` in new files

File: full-service/src/test_utils.rs
Comment (remoun) [inline full-service/src/test_utils.rs:403]: Optional: Restrict this match further with a guard.
```suggestion
            Err(SyncError::Database(WalletDbError::Diesel(
                diesel::result::Error::DatabaseError(kind, info),
            ))) if info.message() == "database is locked" => {
                log::trace!(logger, "Database locked. Will retry"),
                std::thread::sleep(Duration::from_millis(500));
            },
```

**Review by remoun — COMMENTED** [review]
LGTM overall

File: full-service/migrations/2022-03-09-221715_view_only_txos/down.sql
Comment (christian-oudard) [inline full-service/migrations/2022-03-09-221715_view_only_txos/down.sql:2]: Let's have both of these tables in a single migration.

File: full-service/src/json_rpc/wallet.rs
Comment (christian-oudard) [inline full-service/src/json_rpc/wallet.rs:543]: I think that since we're only doing this twice, that we don't need to worry so much about refactoring. If we were doing this multiple times then we should have a helper function. Let's remove the TODO comment.

File: full-service/src/json_rpc/wallet.rs
Comment (christian-oudard) [inline full-service/src/json_rpc/wallet.rs:870]: Let's not leave too many TODO comments.

**Review by christian-oudard — APPROVED** [review]
Approved pending addressing comments.

File: full-service/src/json_rpc/wallet.rs
Comment (Shramp) [reply full-service/src/json_rpc/wallet.rs:870]: my plan was to address these before I merged this PR. They were TODOs for this PR and not TODOs to be left in the code

File: full-service/src/db/view_only_account.rs
Comment (Shramp) [reply full-service/src/db/view_only_account.rs:146]: this makes sense, but do you know why Diesel suggests inlining this import at all? I don't see it mentioned in the documentation and it seems to work when not inlined.

File: full-service/src/json_rpc/wallet.rs
Comment (Shramp) [reply full-service/src/json_rpc/wallet.rs:441]: Tried this and it works, but I'm curious why you prefer it. To me `try_from` seems to make more sense since it's explicitly stating that we're going to create this json_rpc `ViewOnlyAccountSecrets` from our `account` entity. Whereas `try_into` looks more ambiguous.

File: full-service/src/json_rpc/wallet.rs
Comment (Shramp) [reply full-service/src/json_rpc/wallet.rs:872]: yes!

File: full-service/src/db/view_only_account.rs
Comment (david1alvarez) [inline full-service/src/db/view_only_account.rs:20]: Unsure about the `DEFAULT_FIRST_BLOCK_INDEX` value, but the other two look like they should live in mobilecoin/transaction/core/src/constants.rs

File: full-service/src/db/view_only_account.rs
Comment (david1alvarez) [inline full-service/src/db/view_only_account.rs:45]:
```suggestion
        first_block_index: u64,
        import_block_index: u64,
```

File: full-service/src/db/view_only_account.rs
Comment (david1alvarez) [inline full-service/src/db/view_only_account.rs:94]:
```suggestion
        first_block_index: u64,
        import_block_index: u64,
```

File: full-service/src/db/models.rs
Comment (david1alvarez) [inline full-service/src/db/models.rs:121]: each of the `i64` references should be `u64` where negative values don't have meaning

File: full-service/src/db/models.rs
Comment (david1alvarez) [inline full-service/src/db/models.rs:140]: `i64`->`u64`

File: full-service/src/db/view_only_account.rs
Comment (david1alvarez) [inline full-service/src/db/view_only_account.rs:184]: maybe newb question, but why pass `self` here rather than `&self`?

File: full-service/src/db/view_only_txo.rs
Comment (david1alvarez) [inline full-service/src/db/view_only_txo.rs:3]:
```suggestion
//! DB impl for the view-only Txo model.
```

**Review by david1alvarez — COMMENTED** [review]
Generally LGTM, though as I can't comment on the core functionality I'll hold off on officially approving/requesting on my end

File: full-service/src/db/view_only_account.rs
Comment (Shramp) [reply full-service/src/db/view_only_account.rs:23]: I don't want to add these to the mobilecoin repo, I want to keep them in full-service but I agree they should be in a constants file somewhere. Will add that

File: full-service/src/db/models.rs
Comment (Shramp) [reply full-service/src/db/models.rs:121]: I wasn't sure about this. I followed the example of all the other models in our schema, which all use i64. I agree that it would make more sense to use u64 but I think there is something with diesel that makes us use i64

File: full-service/src/db/view_only_account.rs
Comment (Shramp) [reply full-service/src/db/view_only_account.rs:186]: using `self` here instead of `&self` because this is a delete method. I could be wrong, but I think it's unlikely that the entity will be used again after being deleted so it's okay for the method to take ownership

File: full-service/src/json_rpc/balance.rs
Comment (BrianCorbin) [inline full-service/src/json_rpc/balance.rs:78]: So, this will eventually have a field to represent the calculated balance, yeah? Maybe that can wait until we get into implementing tracking spent txos.

File: full-service/src/json_rpc/wallet.rs
Comment (BrianCorbin) [inline full-service/src/json_rpc/wallet.rs:544]: We should order these alphabetically

**Review by BrianCorbin — CHANGES_REQUESTED** [review]

File: full-service/src/json_rpc/balance.rs
Comment (Shramp) [reply full-service/src/json_rpc/balance.rs:78]: yeah, that's my plan

**Review by BrianCorbin — COMMENTED** [review]
LGTM! Just gonna run this on my end a little bit before approving for merge :)

**Review by BrianCorbin — CHANGES_REQUESTED** [review]
the `get_wallet_status` endpoint should probably be updated to include a list of `view_only_accounts` as well. Thoughts on this?

**Review by BrianCorbin — APPROVED** [review]

**Pull request merged**
